hyperspy/io_plugins/nexus.py

```diff
@@ -0,0 +1,1251 @@
+        .e.g search_keys = ["instrument","Fe"] will return
+        hdf entries with instrument or Fe in their hdf path.
+    
+    Returns
+    -------
+    dict
+        When search_list is specified only full paths
+        containing one or more search_keys will be returned 
+        
+    """
+
+    _check_search_keys(search_keys)
+    metadata_dict = {}
+    rootname=""
+
+    # recursive function    
+    def find_searchkeys_in_tree(myDict,rootname):
+        for key, value in myDict.items():
+            if rootname != "":
+                rootkey = rootname +"/"+ key
+            else:
+                rootkey = key
+            if type(search_keys) is list \
+            and any([s1 in rootkey for s1 in search_keys]):
+                mod_keys = _text_split(rootkey, (".","/") )                        
+                # create the key, values in the dict
+                p = metadata_dict
+                for d in mod_keys[:-1]:
+                    p = p.setdefault(d,{})
+                p[mod_keys[-1]] = value
+            if isinstance(value,dict):
+                find_searchkeys_in_tree(value,rootkey)
+                    
+    if search_keys in ["all","hardlinks"]:
+        return tree
+    else:
+        find_searchkeys_in_tree(tree,rootname)
+    return metadata_dict
+
+def _find_smalldata_in_dict(tree):    
+    """Search through a dict and return a dictionary
+    which only contains small metadata (item size < 2)
+    Size of data in attrs groups is ignored
+    This is a convenience method to filter data at time of 
+    saving
+    
+    Parameters
+    ----------
+    tree : dict  
+        input dictionary
+    
+    Returns
+    -------
+    dict
+        When search_list is specified only full paths
+        containing one or more search_keys will be returned 
+        
+    """
+        
+    # recursive function    
+    def find_smallmeta_in_tree(metadata_dict,sizefilter=True):
+        tree={}
+        for key, value in metadata_dict.items():
+            if isinstance(value,(int,float,str,bytes)):
+                tree[key]=value
+            elif isinstance(value,(list,tuple)):
+                if sizefilter:
+                    if len(value) < 2:
+                        tree[key]=value
+                else:
+                    tree[key]=value                    
+            elif isinstance(value,(np.ndarray,da.Array)):
+                if sizefilter:
+                    if value.size < 2:
+                        tree[key]=value
+                else:
+                    tree[key]=value                    
+            elif isinstance(value,dict):
+                if key == "attrs":
+                    tree[key] = find_smallmeta_in_tree(value,sizefilter=False)
+                else:
+                    tree[key] = find_smallmeta_in_tree(value,sizefilter=True)                    
+        return tree
+    metadata_dict=find_smallmeta_in_tree(tree)
+    return metadata_dict
```

Comment:
Same as above: can it be simplified by using the h5py [`visit`](https://h5py.readthedocs.io/en/latest/high/group.html#h5py.Group.visit) function with a nexus specific callable?

Reply:
Based on the comments around lazy etc. I've removed this particular filter to remove data >2 from metadata.
if a user wants to control the metadata they should use the metadata_keys to filter and it simplifies the code.